### PR TITLE
Added DataFetcherFactory support

### PR DESCRIPTION
**Pull request by bbakerman**
Adds a level of indirection into the data fetchers which are specified at
schema def time but needed "wired" at query execution time

see #734

### CODE REVIEW

File: src/main/java/graphql/schema/DataFetcherFactory.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * Creates a {@link graphql.schema.DataFetcherFactory} that always returns the provided {@link graphql.schema.DataFetcher}
+     *
+     * @param dataFetcher the data fetcher to always return
+     * @param <DF>        the type of the data fetcher
+     *
+     * @return a data fetcher factory that always returns the provided data fetcher
+     */
+    static <DF> DataFetcherFactory<DF> useDataFetcher(DataFetcher<DF> dataFetcher) {
```
Comment (andimarek) [inline src/main/java/graphql/schema/DataFetcherFactory.java]: not really sure about that one: Can we the static factory method to somewhere else?

File: src/main/java/graphql/schema/GraphQLFieldDefinition.java [hunk 2]
```diff
@@ -20,27 +20,28 @@
...
+    public GraphQLFieldDefinition(String name, String description, GraphQLOutputType type, DataFetcher<?> dataFetcher, List<GraphQLArgument> arguments, String deprecationReason) {
+        this(name, description, type, DataFetcherFactory.useDataFetcher(dataFetcher), arguments, deprecationReason, null);
     }
 
     @Internal
-    public GraphQLFieldDefinition(String name, String description, GraphQLOutputType type, DataFetcher dataFetcher, List<GraphQLArgument> arguments, String deprecationReason, FieldDefinition definition) {
+    public GraphQLFieldDefinition(String name, String description, GraphQLOutputType type, DataFetcherFactory dataFetcherFactory, List<GraphQLArgument> arguments, String deprecationReason, FieldDefinition definition) {
         assertValidName(name);
-        assertNotNull(dataFetcher, "dataFetcher can't be null");
+        assertNotNull(dataFetcherFactory, "dataFetcherFactory can't be null");
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLFieldDefinition.java]: this is confusing I think: until know we only have `DataFetcher` ... no we say `DataFetcherFactory` can't be null: maybe something like 'you have to provide a DataFetcher (or DataFetcherFactory)' ?

File: src/main/java/graphql/schema/DataFetcherFactory.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public interface DataFetcherFactory<T> {
+
+    /**
+     * Returns a {@link graphql.schema.DataFetcher}
+     *
+     * @param fieldDefinition the {@link graphql.schema.GraphQLFieldDefinition} that needs the data fetcher
+     *
+     * @return a data fetcher
+     */
+    DataFetcher<T> get(GraphQLFieldDefinition fieldDefinition);
```
Comment (andimarek) [inline src/main/java/graphql/schema/DataFetcherFactory.java]: I think it is safer for the future provide an `....Environment` Object containing the `fieldDefitinion`, but also the parent "container".

Comment (bbakerman) [reply src/main/java/graphql/schema/DataFetcherFactory.java]: Great point - on it

File: src/main/java/graphql/schema/GraphQLFieldDefinition.java (hunk 2 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/GraphQLFieldDefinition.java]: done

File: src/main/java/graphql/schema/DataFetcherFactory.java (hunk 1 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/DataFetcherFactory.java]: done

**Pull request merged**